### ISSUE TITLE
Legend is not updated correctly when adding values

### ISSUE THREAD

**Issue by Jerrythafast** [vega/vega#435]
Hey guys,

I just ran into this problem. When you use the Streaming Data API to insert/remove/update data values dynamically, it may sometimes happen that the values in an ordinal scale's domain change. Most of the time, Vega correctly updates the legend to reflect these changes. However, if the number of values in the domain increases beyond the number of values that were in the domain the first time the graph was rendered, the additional values are all rendered at the top of the legend, overlapping each other.

To see this in the Vega Editor, load the [barley](http://vega.github.io/vega-editor/?spec=barley) spec, change all its `enter` property sets to `update`, and then add a data value with a new year:
```js
ved.view.data("barley").insert([{yield: 10, variety: "Trebi", year: 1933, site: "Waseca"}]);
ved.view.update();
```
Notice how the `1933` legend item is rendered in a very odd position.

If we now remove all the `1932` data:
```js
ved.view.data("barley").remove(function(datum){return datum.year == 1932});
ved.view.update();
```
The `1933` legend item jumps (correctly) to where the `1932` legend item used to be. Thus, this problem occurs only if the number of legend items is greater than what it was when first rendering the graph.

Kind regards,
Jerry

**Comment by Jerrythafast** [comment]
Oh, also, perhaps related to this: if the domain of the legend's scale has a `sort` property, it is sorted correctly initially, but any changes to the domain do not take the `sort` property into account. If you remove some values from the domain and then add new ones (possibly the values you just removed), the newly added values simply appear at the bottom of the legend in the order in which they are added. No sorting is done.

**Comment by jheer** [comment]
Thanks Jerry. I took a closer look at this, and as far as I can tell the Legend component is updating properly. The problem appears to be that our GroupBuilder class is maintaining old state and not using an updated scale to drive the legend's internal layout. So, I'm assigning this to @arvind to look at the build pipeline issues.

Here's what I was able to track down: The build/encode/bound pipeline for the legend is initiated at GroupBuilder.js#L263, which uses the current `def` defining the legend component. This invokes the `init` method, which at GroupBuild.js#L37 creates a new `Scale` object. I believe the problem is that this scale is not being updated appropriately. When the legend is reset, the `def` is updated, including a new scale definition for positioning the legend's elements. This new definition appears to be correctly constructed. However, I have not found any code path by which the updated definition is then appropriately re-parsed and the proper scale updates are made. Perhaps when `updateLegend` is invoked (after GroupBuilder.js#L153) additional steps should be taken to update the Builder state?
